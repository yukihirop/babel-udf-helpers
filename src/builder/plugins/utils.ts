import { NodePath } from '@babel/traverse';

// MEMO:
// generate <path> used by path.get(<path>)
export function makePath(path: NodePath): string {
  const parts = [];
  // If the parent exists, set it to path and execute it.Execution ends when there are no parents
  for (; path.parentPath; path = path.parentPath) {
    parts.push(path.key);
    if (path.inList) parts.push(path.listKey);
  }

  return parts.reverse().join('.');
}
