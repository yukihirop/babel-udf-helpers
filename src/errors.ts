class BaseError extends Error {
  // @ts-ignore
  constructor(...params) {
    super(...params);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AlreadyImplementedError);
    }
  }
}

export class AlreadyImplementedError extends BaseError {
  // @ts-ignore
  constructor(...params) {
    super(...params);
    this.name = 'AlreadyImplementedError';
  }
}

export class NotFoundError extends BaseError {
  // @ts-ignore
  constructor(...params) {
    super(...params);
    this.name = 'NotFoundError';
  }
}
