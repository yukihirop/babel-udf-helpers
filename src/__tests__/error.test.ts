import { useDangerousUDFHelpers } from '../';
import { AlreadyImplementedError, NotFoundError } from './../errors';
import { clearAllUDFHelpers } from '../';
import printer from './printer';
import { inputFixturePath } from './utils';

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation();
  jest.spyOn(console, 'warn').mockImplementation();
});

afterEach(() => {
  jest.clearAllMocks();
  clearAllUDFHelpers();
});

describe('addUDFHelper', () => {
  describe('error', () => {
    const type = 'error';

    describe('core_ext', () => {
      describe('babelImplemented', () => {
        // prettier-ignore
        const cases = [
          {
            method: 'BabelFile#addUDFHelper',
            preFunc: (pass, helpers) => {
              try {
                pass.file.constructor.prototype.addUDFHelper = function () { };
                useDangerousUDFHelpers(pass, helpers);
              } catch (e) {
                pass.file.constructor.prototype.addUDFHelper = undefined;
                throw e
              }
            }
          },
          {
            method: 'BabelFile#listUDFHelper',
            preFunc: (pass, helpers) => {
              try {
                pass.file.constructor.prototype.listUDFHelper = function () { };
                useDangerousUDFHelpers(pass, helpers);
              } catch (e) {
                pass.file.constructor.prototype.listUDFHelper = undefined;
                throw e
              }
            }
          },
        ];

        for (const c of cases) {
          const { method, preFunc } = c;
          test(method, () => {
            const dir = 'declaration';
            const helpers = require(inputFixturePath(['basic', dir]));
            const programFuncs = [(pass) => pass.addUDFHelper('declaration')];

            // prettier-ignore
            expect(() => { printer({ helpers, programFuncs, preFunc }); }).toThrowError(new AlreadyImplementedError(
              `unknown:\
 \nThis tool cannot be used. officially supported.
Please see the official documentation.
https://babeljs.io/docs/en/babel-helpers
`));
            // prettier-ignore
            expect(() => { printer({ helpers, programFuncs, preFunc }); }).toThrowErrorMatchingSnapshot();
          });
        }
      });

      test('Not found UDF helpers', () => {
        const programFuncs = [(pass) => pass.addUDFHelper('objectWithoutProperties')];

        // prettier-ignore
        expect(() => { printer({ programFuncs } as any); }).toThrowError(new NotFoundError('unknown: Not found UDF helpers.'));
        // prettier-ignore
        expect(() => { printer({ programFuncs } as any); }).toThrowErrorMatchingSnapshot();
      });

      test('babelAlreadyDefined', () => {
        const dir = 'babelAlreadyDefined';
        const helpers = require(inputFixturePath([type, dir]));
        const programFuncs = [(pass) => pass.addUDFHelper('objectWithoutProperties')];

        // prettier-ignore
        expect(() => { printer({ helpers, programFuncs }); }).toThrowError(new AlreadyImplementedError('unknown: objectWithoutProperties cannot be used because it is supported by babel official.\nPlease change the name of the helper.'));
        // prettier-ignore
        expect(() => { printer({ helpers, programFuncs }); }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('DependencyResolvePlugin', () => {
      // prettier-ignore
      const cases = [
        { dir: 'doNotExist', err: new ReferenceError('unknown: Unknown helper doNotExist') },
        { dir: 'importNamed', err: new SyntaxError('unknown: UDF Helpers can only import a default value (This is an error on an internal node. Probably an internal error.)') },
        { dir: 'doNotGiveName', err: new SyntaxError('unknown: UDF Helpers should give names to their exported func declaration (This is an error on an internal node. Probably an internal error.)') },
        { dir: 'expression', err: new SyntaxError('unknown: UDF Helpers must be a function declaration (This is an error on an internal node. Probably an internal error.)') },
        { dir: 'exportAll', err: new SyntaxError('unknown: UDF Helpers can only export default (This is an error on an internal node. Probably an internal error.)') },
        { dir: 'exportNamed', err: new SyntaxError('unknown: UDF Helpers can only export default (This is an error on an internal node. Probably an internal error.)') },
        { dir: 'doNotExportDefault', err: new SyntaxError('unknown: UDF Helpers must default-export something.') },
      ];

      for (const c of cases) {
        const { dir, err } = c;

        test(dir, () => {
          const helpers = require(inputFixturePath([type, dir]));
          const programFuncs = [(pass) => pass.addUDFHelper(dir)];

          // prettier-ignore
          expect(() => { printer({ helpers, programFuncs }); }).toThrowError(err);
          // prettier-ignore
          expect(() => { printer({ helpers, programFuncs }); }).toThrowErrorMatchingSnapshot();
        });
      }
    });
  });
});
