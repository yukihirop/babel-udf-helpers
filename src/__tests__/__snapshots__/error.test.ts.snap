// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUDFHelper error DependencyResolvePlugin doNotExist 1`] = `"unknown: Unknown helper doNotExist"`;

exports[`addUDFHelper error DependencyResolvePlugin doNotExportDefault 1`] = `"unknown: UDF Helpers must default-export something."`;

exports[`addUDFHelper error DependencyResolvePlugin doNotGiveName 1`] = `"unknown: UDF Helpers should give names to their exported func declaration (This is an error on an internal node. Probably an internal error.)"`;

exports[`addUDFHelper error DependencyResolvePlugin exportAll 1`] = `"unknown: UDF Helpers can only export default (This is an error on an internal node. Probably an internal error.)"`;

exports[`addUDFHelper error DependencyResolvePlugin exportNamed 1`] = `"unknown: UDF Helpers can only export default (This is an error on an internal node. Probably an internal error.)"`;

exports[`addUDFHelper error DependencyResolvePlugin expression 1`] = `"unknown: UDF Helpers must be a function declaration (This is an error on an internal node. Probably an internal error.)"`;

exports[`addUDFHelper error DependencyResolvePlugin importNamed 1`] = `"unknown: UDF Helpers can only import a default value (This is an error on an internal node. Probably an internal error.)"`;

exports[`addUDFHelper error core_ext Not found UDF helpers 1`] = `"unknown: Not found UDF helpers."`;

exports[`addUDFHelper error core_ext babelAlreadyDefined 1`] = `
"unknown: objectWithoutProperties cannot be used because it is supported by babel official.
Please change the name of the helper."
`;

exports[`addUDFHelper error core_ext babelImplemented BabelFile#addUDFHelper 1`] = `
"unknown: 
This tool cannot be used. officially supported.
Please see the official documentation.
https://babeljs.io/docs/en/babel-helpers
"
`;

exports[`addUDFHelper error core_ext babelImplemented BabelFile#listUDFHelper 1`] = `
"unknown: 
This tool cannot be used. officially supported.
Please see the official documentation.
https://babeljs.io/docs/en/babel-helpers
"
`;
